sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -120,118 +123,138 @@ cdef float compute_gradient_positive(float[:] val_P,
         int ax
         long i, j, k
         long n_samples = indptr.shape[0] - 1
-        float dij, qij, pij
         float C = 0.0
+        float dij, qij, pij
         float exponent = (dof + 1.0) / 2.0
         float float_dof = (float) (dof)
-        float[3] buff
+        float* buff
```

Comment:
Question: why did you go from array to pointer + malloc?

At first sight I don't see the benefit, especially considering that the call to malloc is longer than just allocating the array on the stack

Reply:
we need a buffer array in each thread. This way buff is a thread local buffer